### ISSUE TITLE
New Rule Proposal: no-use-object-member-before-assign

### ISSUE THREAD

**Issue by stevendesu** [eslint/eslint#11588]
**Please describe what the rule should do:**

To the best of ESLint's ability, it should detect if an object member _*might*_ be defined before allowing it to be used (as an operand for any operator)

There are some times that it will be admittedly impossible to know if an object member is defined (object members can be set via user input), but there are a few cases that should be easy to catch.

In general an object's member existence can be determined with the following formula:

 - If an object was just created:
     - If any members are declared using ES6's variable key syntax (`var obj = { [a]: 3 };`) then the object is in "blacklist" mode -- any members ***may*** exist unless explicitly deleted
     - If ***all*** members are declared statically (or no members are declared) then the object is in "whitelist" mode -- ***no members*** exist unless explicitly added
 - In whitelist mode:
     - If the dot (`.`) operator or bracket (`[]`) operator are used to set a **static** member on an object (e.g. `obj["a"] = 3`), only that static member is added to the whitelist
     - If the delete (`delete`) operator is used to unset a static member on an object (e.g. `delete obj.a`), then that member is removed from the whitelist
     - If the bracket (`[]`) operator is used to set a variable member on an object (e.g. `obj[a] = 3`), then the object is converted to "blacklist" mode. If it was formerly in "whitelist" mode, the whitelist is cleared
    - If the bracket (`[]`) operator and delete (`delete`) operator are used to delete a variable member on an object (e.g. `delete obj[a]`), then do nothing (we cannot know which element was deleted)
 - In blacklist mode:
     - If the dot (`.`) operator or bracket (`[]`) operator are used to set a **static** member on an object (e.g. `obj["a"] = 3`), then that member is removed from the blacklist
     - If the delete (`delete`) operator is used to unset a static member on an object (e.g. `delete obj.a`) then that member is added to the blacklist
    - If the bracket (`[]`) operator is used to set a variable member on an object (e.g. `obj[a] = 3`), then the blacklist is cleared
    - If the bracket (`[]`) operator and delete (`delete`) operator are used to delete a variable member on an object (e.g. `delete obj[a]`), then do nothing (we cannot know which element was deleted)

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

First, what counts as "using" an object member:
```js
var obj = {}:

// Math operators:
obj.a + obj.b - obj.c * obj.d / obj.e;

// Member access operators (not on the object, but on its members):
obj.a.b;
obj.a[b];

// Function call operator:
obj.a();

// Bitwise operators:
obj.a >> obj.b << obj.c | obj.d & obj.e ^ ~obj.f;

// Right-hand side of assignment operator
var a = obj.a;

// Second or third operand of ternary operator
true ? obj.a : obj.b;
```

The following do NOT count as "using" an object member:
```js
// Left-hand side of assignment operator
obj.a = 3;

// Stand-alone expression
obj.a;
if (obj.a) { /* do stuff */ }

// Typeof operator
typeof obj.a;

// First operand of ternary operator
obj.a ? true : false;
```

Now examples of using object members that are obviously undefined:
```js
// Using an object member before any have been assigned
var obj = {};
obj.a();

// Using an object member after only assignments are static
var obj = {
    b: 1
};
obj.c = 2;
obj["d"] = 3;
obj.a();

// Using an object member after deletion
var obj = {
    a: function() { console.log("I exist!"); }
};
delete obj.a;
obj.a();

// A simple typo which would lead to a currently uncaught bug
var student = {
    name: "Steven",
    age: 28,
    gpa: 2.0
};
console.log("Hello, " + student.nmae); // Hello, undefined
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is an extension to the ESLint rule [no-use-before-define](https://eslint.org/docs/rules/no-use-before-define) to try and catch typos and bugs before they crop up

It's clearly not ***perfect***, because variable member names allow for unknown existence:

```js
var obj = {};
var key = prompt("Enter a key:");
obj[key] = 1;
console.log( obj["steven"] ); // May be undefined, may be 1
```

However it will at least catch typos, variables that were deleted without updating other parts of the code, renamed variables, and other common coding mistakes.

**Are you willing to submit a pull request to implement this rule?**

I'd be willing, but it would take quite some time to learn how ESLint works

**Comment by platinumazure** [comment]
Hi @stevendesu, thanks for the issue.

I could see this being vaguely possible using code path analysis, but it could be a pretty complex rule.

I'll also note that using something like Typescript would allow you to get this for free. :smile:

**Comment by stevendesu** [comment]
:+1: on the TypeScript suggestion -- I should definitely start using that. So far I've been using ESLint to avoid making the switch to TypeScript :smile: I get most of the same benefits (plus some unique benefits) and I don't have to add a build step -- I can run my code exactly as I type it. Which is really nice when rapidly prototyping small units (I don't have to run Webpack on my entire project to test out one new class I wrote)

Still, I think making the change to TypeScript is long overdue.
